[4.0] Decoupling highlighter plugin from com_finder

Hackwar opened the pull request:
This PR tries to fix #20252. 

This PR introduces a new event, "onFinderResult", which takes a single result object and the search query object. This allows plugins to modify the results (for example changing MIME types, link types, adding additional data, etc.) and in this case it allows the highlight plugin to modify the URL of the link of the search result to add the highlight parameter. This decouples the highlight system plugin from com_finder.

~~~The downside in this case is, that the URL, that is displayed below the search result, now also contains the highlight parameter. I would say that we should rework the output of the results anyway, but that would be the topic of another PR.~~~
The ```FinderIndexerResult``` object has both an ```url``` and ```route``` attribute and we are now using ```url``` for display purposes and ```route``` for the links href.
I also tried to change a few of the calls to the new namespaced versions.

[UPDATE]
In further reviews, I also found out the issue why searchwords with utf8 characters are not properly highlighted. It turns out that OutputFilter::stringJSSafe() is not utf8 aware and I had to rewrite it. Now it properly encodes this into a utf8 escaped string.

[UPDATE]
Yet another update regarding the ```url``` and ```route```: url is actually the ID of the item and thus can not be used. Instead I added ```cleanurl``` instead.

## How to test
* Test smart search and see the highlighting parameter in the URL of the search results.
* Apply the patch and see that it works as before, except that the additional highlight parameter is in the URL below the result.
* This PR mainly needs a codereview.

Hackwar committed:
Decoupling highlighter plugin from com_finder

Hackwar committed:
Codestyle

Quy commented on components/com_finder/View/Search/HtmlView.php:
```diff
@@ -112,19 +116,19 @@ class HtmlView extends BaseHtmlView
 	{
-		$app    = \JFactory::getApplication();
-		$params = $app->getParams();
+		$app    = Factory::getApplication();
```
Remove extra spaces before `=` or align with `=`.

carlitorweb commented on plugins/system/highlight/highlight.php:
```diff
@@ -88,4 +88,32 @@ public function onAfterDispatch()
+	 *
+	 * @return  void
+	 *
+	 * @since   4.0
```
change to `__DEPLOY_VERSION__`

Hackwar committed:
Codestyle

Hackwar committed:
Version comment

carlitorweb commented:
I got this. This PR depend of other?
![screenshot_20180524154204](https://user-images.githubusercontent.com/3956472/40509879-d223dff0-5f68-11e8-8944-90f7a9ed7d2f.png)

carlitorweb commented:
I got this. This PR depend of other?
![error-1055--joomla-4devtparent-id--isn-t-in-group-by---google-chrome](https://user-images.githubusercontent.com/3956472/40509998-40f22ed2-5f69-11e8-8d85-cefb0c822027.png)

Hackwar commented:
No, but there is another bug that is fixed in #20185. Looks like this strict query handling was disabled again in a later version of 4.0-dev. Idk. Maybe update your 4.0-dev branch to the current state? It works there...

carlitorweb commented:
I test now without this PR, and give the same error. So, is not this PR.  And yes, the branch is up to the current state

carlitorweb commented:
Ok, the error I get is fixed here #20185. Beside code review will be good test this, but we need #20185 merged for that

Quy commented on plugins/system/highlight/highlight.php:
```diff
@@ -88,4 +88,32 @@ public function onAfterDispatch()
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function onFinderResult($item, $query)
```
There is `afterFinderResults`. For consistency, make it `onFinderResults` with a `s`?

Hackwar replied:
Yes, there is a ```afterFinderResults```, but that is only the name of a mark in the logging, not a plugin event. Since I want to run the event on each result separately, I don't really would want to add the ```s``` to the event name...

Hackwar committed:
Merge branch '4.0-dev' of https://github.com/joomla/joomla-cms into j4finder_highlighter

# Conflicts:
#	components/com_finder/View/Search/HtmlView.php

Hackwar commented:
#20185 has been merged in here. Happy testing. :smile:

carlitorweb commented:
Got this:
 `Warning: Invalid argument supplied for foreach() in .....components\com_finder\tmpl\search\default_results.php on line 49`

carlitorweb commented on components/com_finder/View/Search/HtmlView.php:
```diff
@@ -112,19 +116,19 @@ class HtmlView extends BaseHtmlView
+		$this->state = $this->get('State');
+		$this->query = $this->get('Query');
+		\JDEBUG ? Profiler::getInstance('Application')->mark('afterFinderQuery') : null;
+		$this->results = $this->get('Results');
```
Revert this one, is  `$this->get('Items') `

carlitorweb commented:
All work ok. Waiting for the fix of the error

Hackwar committed:
Fixing bug due to #20185

Hackwar commented:
This was a bug because of a conflict between the original code and #20185. In #20185 it was changed to "Items" and somehow that change was overwritten here. Should work now.

carlitorweb commented:
I have tested this item :white_check_mark: successfully on a52bd581ad5ab0a0aba8cc36e6e024572a566da0<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20571">issues.joomla.org/tracker/joomla-cms/20571</a>.</sub>

Hackwar committed:
Using url for display and route for route-URL

Hackwar committed:
Merge branch '4.0-dev' into j4finder_highlighter

Hackwar commented:
I've updated the PR and fixed some things. It would be nice if you could test this again and if we could get one more tester for this. :smile:

Quy commented:
The URLs don't match where the only difference should be the highlight query string.

```
http://localhost/joomla-cms-j4finder_highlighter/index.php/blog/5-your-modules?highlight=WyJqb29tbGEiXQ==

http://localhost/joomla-cms-j4finder_highlighter/index.php/component/content/article/5-your-modules
```

Hackwar commented:
The issue here is, that ```FinderIndexerAdapter::getURL()``` returns the wrong URL. That method should actually be deprecated... We have to fix the finder plugins for that.

Quy commented on components/com_finder/tmpl/search/default_result.php:
```diff
@@ -35,22 +35,11 @@
-	$route .= '&highlight=' . base64_encode(json_encode($this->query->highlight));
-}
-
+$url = JRoute::_($this->result->route);
```
Add a blank line above. Or not assign it to a variable and leave it where it was before.

Hackwar committed:
Merge branch '4.0-dev' into j4finder_highlighter

Hackwar commented:
I've changed the way the URLs are produced, which should fix the issue that you had and also cleaned up the result layout. Please test again.

Hackwar commented:
Please delay that. I made a mistake. I will report back when I fixed the issue.

Hackwar committed:
Cleaning up finder plugins and adding cleanURL attribute in model

Hackwar commented:
Ok, I got this right now. I removed the FinderIndexerHelper::getContentPath() method, since we don't need that with the new router anymore. I also cleaned up the finder plugins and improved the comments there. Last but not least, I added a ```cleanURL``` attribute after unserialising the object in the model, where I simply copy the ```route``` attribute to this. That allows us to render the clean URL in the result layout and at the same time allow the highlighter plugin to add to the URL.

So please test and have a go. :smile:

Hackwar committed:
Making OutputFilter::stringJSSafe() utf8 aware

Hackwar commented:
Since @infograf768 requested this, I also reviewed the highlighting of search words with utf8 chars in the search results page. It turns out that \Joomla\CMS\Filter\OutputFilter::stringJSSafe() is not utf8 aware and mangles all characters that are not ASCII. I rewrote that and now it should work fine for utf8. Looking for your test, @infograf768 :wink:

Hackwar commented:
The suggestions have nothing to do with highlighting. I hope no one now gets confused and delays this PR because you are bringing up unrelated bugs here.

infograf768 commented:
I sincereley hope not. As usual I test a few things when I test any of your PRs. Just consider it as a note about something to remember if it can be solved or not.

As I only tested the highlighting and not the hardcoupling issue, I can't set the my test as OK on issues.joomla.org

carlitorweb commented:
About search for specific phrases, we can continue this talk in #20384 , is where @Hackwar change this.

carlitorweb commented:
I have tested this item :white_check_mark: successfully on 9834dd9087c45f035470ccacca6ee61054c6f4f2<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20571">issues.joomla.org/tracker/joomla-cms/20571</a>.</sub>

infograf768 commented:
I have tested this item :white_check_mark: successfully on 9834dd9087c45f035470ccacca6ee61054c6f4f2<br /><br />@laoneo @mbabker 
This PR is fine concerning the highlighting of terms. But I have no idea about the decoupling.

Can you have a look at that aspect and set RTC/Merge if OK?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20571">issues.joomla.org/tracker/joomla-cms/20571</a>.</sub>

laoneo commented on plugins/finder/categories/categories.php:
```diff
@@ -304,8 +306,6 @@ protected function index(FinderIndexerResult $item, $format = 'html')
 			$item->route = ContentHelperRoute::getCategoryRoute($item->id, $item->language);
 		}
 
-		$item->path = FinderIndexerHelper::getContentPath($item->route);
```
Is this removal necessary then or can J3 plugins still work with that code in place?

Hackwar commented:
Since this is more or less broken code and potentially creates wrong URLs, this code will not work in 4.0 anymore. Plugins from 3.x can't use it anymore and shouldn't use it now already.Am 29.06.2018 07:56 schrieb Allon Moritz <notifications@github.com>:@laoneo commented on this pull request.



In plugins/finder/categories/categories.php:
> @@ -304,8 +306,6 @@ protected function index(FinderIndexerResult $item, $format = 'html')
 			$item->route = ContentHelperRoute::getCategoryRoute($item->id, $item->language);
 		}
 
-		$item->path = FinderIndexerHelper::getContentPath($item->route);

Is this removal necessary then or can J3 plugins still work with that code in place?

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.

Hackwar commented:
I'm going to create a depreciation PR soon.Am 29.06.2018 09:26 schrieb Allon Moritz <notifications@github.com>:Is it then deprecated in J3?

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.

Hackwar commented:
PR for deprecation: #20934

Hackwar committed:
Merge branch '4.0-dev' into j4finder_highlighter

Hackwar committed:
Merge branch '4.0-dev' into j4finder_highlighter

carlitorweb commented:
I have tested this item :white_check_mark: successfully on 55057f220f66402c3162024d617cdbbe945bacc8<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20571">issues.joomla.org/tracker/joomla-cms/20571</a>.</sub>

infograf768 commented:
I have tested this item :white_check_mark: successfully on 55057f220f66402c3162024d617cdbbe945bacc8<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20571">issues.joomla.org/tracker/joomla-cms/20571</a>.</sub>

infograf768 commented:
RTC. Thanks<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/20571">issues.joomla.org/tracker/joomla-cms/20571</a>.</sub>

Hackwar commented:
Thank you!

laoneo committed:
Merge branch '4.0-dev' into j4finder_highlighter
